cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -1018,30 +1006,33 @@ runQueryStakePools
   -> ExceptT ShelleyQueryCmdError IO ()
 runQueryStakePools (AnyConsensusModeParams cModeParams)
                           network mOutFile = do
-  SocketPath sockPath <- firstExceptT ShelleyQueryCmdEnvVarSocketErr
-                           $ newExceptT readEnvSocketPath
+  SocketPath sockPath <- lift readEnvSocketPath & onLeft (throwE . ShelleyQueryCmdEnvVarSocketErr)
 
   let localNodeConnInfo = LocalNodeConnectInfo cModeParams network sockPath
 
-  result <- ExceptT . fmap (join . first ShelleyQueryCmdAcquireFailure) $
-    executeLocalStateQueryExpr localNodeConnInfo Nothing $ runExceptT @ShelleyQueryCmdError $ do
-      anyE@(AnyCardanoEra era) <- case consensusModeOnly cModeParams of
-        ByronMode -> return $ AnyCardanoEra ByronEra
-        ShelleyMode -> return $ AnyCardanoEra ShelleyEra
-        CardanoMode -> lift . queryExpr $ QueryCurrentEra CardanoModeIsMultiEra
+  poolIds <-
+    ( lift $ executeLocalStateQueryExpr localNodeConnInfo Nothing $ runExceptT @ShelleyQueryCmdError $ do
+        anyE@(AnyCardanoEra era) <- case consensusModeOnly cModeParams of
+          ByronMode -> return $ AnyCardanoEra ByronEra
+          ShelleyMode -> return $ AnyCardanoEra ShelleyEra
+          CardanoMode ->
+            lift (queryExpr $ QueryCurrentEra CardanoModeIsMultiEra)
+              & onLeft (throwE . ShelleyQueryCmdUnsupportedNtcVersion)
 
-      let cMode = consensusModeOnly cModeParams
+        let cMode = consensusModeOnly cModeParams
 
-      case toEraInMode era cMode of
-        Just eInMode -> do
-          sbe <- getSbe $ cardanoEraStyle era
+        eInMode <- toEraInMode era cMode
+          & hoistMaybe (ShelleyQueryCmdEraConsensusModeMismatch (AnyConsensusMode cMode) anyE)
```

Comment:
This can also be written as:

```
        eInMode <- pure (toEraInMode era cMode)
          & onNothing (throwE $ ShelleyQueryCmdEraConsensusModeMismatch (AnyConsensusMode cMode) anyE)
```

This shows the versatility of the `on*` combinators.